test/forcing/CsvPerFeatureForcingProvider_Test.cpp

```diff
@@ -164,3 +176,63 @@ TEST_F(CsvPerFeatureForcingProviderTest, TestGetAvailableForcingOutputs)
 
 }
 
+///Test CSV Units Parsing
+TEST_F(CsvPerFeatureForcingProviderTest, TestForcingUnitHeaderParsing)
+{
+    time_t begin = Forcing_Object_3->get_data_start_time();
+    int i = 8;
+    time_t t = begin+(i*3600);
+
+    const std::vector<std::string>& varnames = this->Forcing_Object_3->get_avaliable_variable_names();
+    const std::vector<std::string>& expected = {
+        "RAINRATE",
+        "T2D",
+        "Q2D",
+        "U2D",
+        "V2D",
+        "TEST",
+        "PSFC[Pa)",
+        "SWDOWN(W m-2]",
+        "LWDOWN      (W m-2]"
+    };
+
+    for (auto ite = expected.begin(); ite != expected.end(); ite++) {
+        // make sure each expected column name is within varnames
+        EXPECT_NE(std::find(varnames.begin(), varnames.end(), *ite), varnames.end());
+
+        testing::internal::CaptureStderr();
+        const auto val = this->Forcing_Object_3->get_value(
+            CSVDataSelector(*ite, t, 3600, ""),
+            data_access::SUM
+        );
+        const std::string cerr_output = testing::internal::GetCapturedStderr();
+
+        if (ite - expected.begin() < 6) {
+            // we are expecting warnings, since udunits is trying to map some unit to no units
+            // if we don't get a warning, then the unit type was not parsed
+            EXPECT_NE(cerr_output, "");
```

Comment:
I agree... I'm not a fan of the way I implemented these cases haha.

One way that I think would be better is to expose a getter on `CsvPerFeatureForcingProvider` that returns a read-only reference to the `available_forcings_unit` map. Then, I can test for expected units directly. How does that sound?

Reply:
If we had such a method it would probably belong in `DataProvider` though... that may be a good idea or may be something we don't want to do--in any case then you'd have to implement it in `NetCDFPerFeatureDataProvider` also... not hard, but scope creep.

How about, since we know the units that it's supposed to be, we go ahead and provide a valid unit for conversion to `CSVDataSelector` for cases 0-5 and what _would_ be a valid unit for conversion for cases 6-8, then check if the result of `get_value` changes (in case 0-5 it should and in cases 6-8 it should not)? Make sense? Flaw?

Reply:
See commit 729e70a, this implements your suggestion. I've also removed the unit conversion checks at the end of this unit test, since they are redundant with what's tested in `TestForcingDataUnitConversion`. Thanks!